### PR TITLE
[ISSUE-246][REFACTOR] Refactor LifecycleManager to make it's code more clear and more readable

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### How was this patch tested?

/cc @related-reviewer

/assign @main-reviewer

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @waitinfuture

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 1]
```diff
@@ -220,21 +220,22 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
-    applicationId: String,
-    shuffleId: Int,
-    numMappers: Int,
-    numPartitions: Int): Unit = {
-    // check if same request already exists for the same shuffle.
-    // If do, just register and return
-    registerShuffleRequest.synchronized {
-      if (registerShuffleRequest.containsKey(shuffleId)) {
+  private def handleRegisterShuffle(
+      context: RpcCallContext,
```
[the remark this answers is not in the document]
Comment (FMX) [reply client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:224]: I think it`s okay. I'll change all indentations recently.

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 2]
```diff
@@ -434,48 +437,55 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
+
     val candidates = workersNotBlacklisted(shuffleId)
-    val slots = if (oldPartition != null) {
+    if (candidates.size < 1 || (ShouldReplicate && candidates.size < 2)) {
+      logError("[Update partition] failed for not enough candidates for revive.")
+      reply(ChangeLocationResponse(StatusCode.SlotNotAvailable, null))
+      return null
+    }
+
+    val revivedLocation = if (oldPartition != null) {
```
Comment (FMX) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:454]: This slot can be revived or split.  There is no reason to assume this slot is intended for reviving context.

File: client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala [hunk 3]
```diff
@@ -992,65 +1047,65 @@ class LifecycleManager(appId: String, val conf: RssConf) extends RpcEndpoint wit
...
         res = requestDestroy(workerInfo.endpoint,
           Destroy(shuffleKey, res.failedMasters, res.failedSlaves))
       }
-      failedMasters.addAll(res.failedMasters)
-      failedSlaves.addAll(res.failedSlaves)
+      failedMasterLocationUniqueIds.addAll(res.failedMasters)
+      failedSlaveLocationUniqueIds.addAll(res.failedSlaves)
     }
-    (failedMasters, failedSlaves)
+    (failedMasterLocationUniqueIds, failedSlaveLocationUniqueIds)
```
Comment (FMX) [inline client/src/main/scala/com/aliyun/emr/rss/client/write/LifecycleManager.scala:1108]: We ignored the result of the destroyed message. So I think we can remove the result list.

**Review by FMX — COMMENTED** [review]
This patch surely improved readability.

**Review by FMX — APPROVED** [review]
LGTM.

**Comment by waitinfuture** [comment]
I will run a e2e test.

**Pull request merged**